flowmachine/flowmachine/core/server/action_handlers.py

```diff
@@ -0,0 +1,327 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#
+# This module contains the handler functions for the actions which are
+# supported by the flowmachine server. In order to implement a new action
+# you need to provide a handler function (see below for some examples) and
+# register it in the ACTION_HANDLERS lookup. The return value of the action
+# handler must be of type `ZMQReply`.
+#
+# In order to actually perform an action, call the `perform_action()`
+# function with the action name and parameters. This selects the correct
+# action handler and also gracefully handles any potential errors.
+#
+
+from apispec import APISpec
+from apispec_oneofschema import MarshmallowPlugin
+from marshmallow import ValidationError
+
+from flowmachine.core import Query, GeoTable
+from flowmachine.core.cache import get_query_object_by_id
+from flowmachine.core.query_info_lookup import (
+    QueryInfoLookup,
+    UnkownQueryIdError,
+    QueryInfoLookupError,
+)
+from flowmachine.core.query_state import QueryStateMachine, QueryState
+from flowmachine.utils import convert_dict_keys_to_strings, sort_recursively
+from .exceptions import FlowmachineServerError
+from .query_schemas import FlowmachineQuerySchema
+from .zmq_helpers import ZMQReply
+
+__all__ = ["perform_action"]
+
+
+def action_handler__ping():
+    """
+    Handler for 'ping' action.
+
+    Returns the message 'pong'.
+    """
+    return ZMQReply(status="success", msg="pong")
+
+
+def action_handler__get_available_queries():
+    """
+    Handler for 'get_available_queries' action.
+
+    Returns a list of available flowmachine queries.
+    """
+    available_queries = list(FlowmachineQuerySchema.type_schemas.keys())
+    return ZMQReply(status="success", payload={"available_queries": available_queries})
+
+
+def action_handler__get_query_schemas():
+    """
+    Handler for the 'get_query_schemas' action.
+
+    Returns a dict with all supported flowmachine queries as keys
+    and the associated schema for the query parameters as values.
+    """
+    spec = APISpec(
+        title="FlowAPI",
+        version="1.0.0",
+        openapi_version="3.0.2",
+        plugins=[MarshmallowPlugin()],
+    )
+    spec.components.schema("FlowmachineQuerySchema", schema=FlowmachineQuerySchema)
+    schemas_spec = spec.to_dict()["components"]["schemas"]
+    schemas_spec = sort_recursively(schemas_spec)  # needed for reproducible output
```

Comment:
The reproducibility part is significant for the tests, right? Wouldn't we rather sort in the test and avoid paying for that here?

Reply:
Yes, good point. It's also kind of nice to return the schemas in the order they are defined internally.